File: src/task_management/service.py
import uuid
from typing import Optional
from .repository import Repository, Task, Message

class TaskService:
    """Service for managing tasks"""
    
    def __init__(self, repository: Repository):
        self.repository = repository
    
    async def create_task(self, content: str, task_id: Optional[str] = None) -> Task:
        """Create a new task with the given content"""
        if not task_id:
            task_id = str(uuid.uuid4())
            
        task = Task(
            id=task_id,
            messages=[Message(type="user", content=content)],
            status="PENDING"
        )
        
        try:
            return await self.repository.create_task(task)
        except ValueError:
            # If task exists, get it and append the new message
            existing_task = await self.repository.get_task(task_id)
            existing_task.messages.append(Message(type="user", content=content))
            return await self.repository.update_task(existing_task)
    
    async def process_task(self, task_id: str, response: str) -> Task:
        """Process a task with the given response"""
        task = await self.repository.get_task(task_id)
        task.messages.append(Message(type="assistant", content=response))
        task.status = "COMPLETED"
        return await self.repository.update_task(task)
    
    async def get_task(self, task_id: str) -> Task:
        """Get a task by ID"""
        return await self.repository.get_task(task_id)
    
    async def delete_task(self, task_id: str) -> None:
        """Delete a task by ID"""
        await self.repository.delete_task(task_id) 

File: gateways/pkg/board/service.go
package board

import (
	"context"
	"errors"
	"log/slog"

	"github.com/gofrs/uuid/v5"
	"gitlab.skypicker.com/platform/experimental/agents/gateways/pkg/board/repository"
	"gitlab.skypicker.com/platform/experimental/agents/gateways/pkg/genkit"
)

type Service struct {
	repo      repository.Repository
	genkitSvc *genkit.Service
}

type Request struct {
	Task repository.Task
}
type Response struct {
	Task repository.Task
}

func NewService(repo repository.Repository, genkitSvc *genkit.Service) *Service {
	return &Service{repo: repo, genkitSvc: genkitSvc}
}

func (s *Service) CreateTask(ctx context.Context, req Request) (Response, error) {
	if req.Task.ID == "" {
		id, err := uuid.NewV4()
		if err != nil {
			return Response{}, err
		}
		req.Task.ID = id.String()
	}
	task, err := s.repo.CreateTask(ctx, req.Task)
	if err != nil {
		task, err = s.repo.GetTask(ctx, req.Task.ID)
		if err != nil {
			return Response{}, err
		}
		task.Messages = append(task.Messages, repository.Message{
			Content: req.Task.Messages[len(req.Task.Messages)-1].Content,
			Type:    "user",
		})
	}
	if len(task.Messages) == 0 {
		return Response{}, errors.New("task has no messages")
	}
	res, err := s.genkitSvc.SendRequest(ctx, &genkit.Request{
		Data: genkit.RequestData{
			SessionKey: task.ID,
			Message:    task.Messages[len(task.Messages)-1].Content,
		},
	})
	if err != nil {
		slog.Error("error sending request to genkit", "error", err)
		task.Status = "ERROR"
	} else {
		task.Messages = append(task.Messages, repository.Message{
			Content: res.Result,
			Type:    "assistant",
		})
		task.Status = "OK"
	}
	if task, err = s.repo.UpdateTask(ctx, task); err != nil {
		return Response{}, err
	}

	return Response{
		Task: task,
	}, nil
}
